dockstore-webservice/src/main/java/io/dockstore/webservice/languages/IpynbHandler.java

```diff
@@ -0,0 +1,318 @@
+        if (!file.isPresent()) {
+            return negativeValidation(notebookPath, "No notebook file is present");
+        }
+        String content = file.get().getContent();
+
+        // Parse the notebook.
+        Nbformat notebook;
+        try {
+            notebook = parseNotebook(content);
+        } catch (JsonParseException ex) {
+            return negativeValidation(notebookPath, "The notebook file is malformed", ex);
+        }
+
+        // Confirm that the entry's programming language (descriptor type subclass) matches the notebook's programming language.
+        try {
+            String entryLanguage = workflow.getDescriptorTypeSubclass().toString();
+            String notebookLanguage = extractProgrammingLanguage(notebook);
+            if (!entryLanguage.equalsIgnoreCase(notebookLanguage)) {
+                return negativeValidation(notebookPath, String.format("The notebook programming language must be '%s'", entryLanguage));
+            }
+        } catch (JsonParseException ex) {
+            return negativeValidation(notebookPath, "Error reading the notebook programming language", ex);
+        }
+
+        // If we reach this point, everything is well, return a positive validation.
+        return positiveValidation();
+    }
+
+    private String extractProgrammingLanguage(Nbformat notebook) {
+        return notebook.getMetadata().optLanguageInfo().flatMap(Nbformat.Metadata.LanguageInfo::optName).orElse(PYTHON);
+    }
+
+    @Override
+    public VersionTypeValidation validateToolSet(Set<SourceFile> sourceFiles, String primaryDescriptorFilePath) {
+        throw new UnsupportedOperationException("Notebooks do not support tools");
+    }
+
+    @Override
+    public VersionTypeValidation validateTestParameterSet(Set<SourceFile> sourceFiles) {
+        // For now, until we determine the type(s) of test files we should be validating, return without inspecting the files.
+        return positiveValidation();
+    }
+
+    private VersionTypeValidation positiveValidation() {
+        LOG.info("Created positive validation");
+        return new VersionTypeValidation(true, Collections.emptyMap());
+    }
+
+    private VersionTypeValidation negativeValidation(String path, String message) {
+        LOG.warn("Created negative validation, file {}: {}", path, message);
+        return new VersionTypeValidation(false, Map.of(path, message));
+    }
+
+    private VersionTypeValidation negativeValidation(String path, String message, Exception ex) {
+        String reason = ex.getMessage();
+        if (reason != null) {
+            message = String.format("%s: %s", message, reason);
+        }
+        LOG.warn("Created negative validation, file {}: {}", path, message, ex);
+        return new VersionTypeValidation(false, Map.of(path, message));
+    }
+
+    /**
+     * Partial read-only POJO representation of a parsed notebook file.
+     */
+    public static class Nbformat {
```

Comment:
Kinda wonder if this should be in `IpynbHandler` and whether the handler for different kinds of notebooks would be relatively similar (or not!) with a different POJO representation (or not). 

I guess we can wait until we have more examples before generalizing and or using the plugin language system.